### PR TITLE
VSI performance checks

### PR DESCRIPTION
**Pull request by quantumsteve**
Description of work.

This applies changes suggested by the [clang-tidy performance-* checks](http://clang.llvm.org/extra/clang-tidy/checks/list.html). Reviewing these changes, I found a few more places to reduce the number of unnecessary copies.  

I don't think these changes are in hot loops that would be obvious to end users. 

**To test:**

Review code changes and builds status. If you notice a significant performance improvement, let's add it to the release notes.

There is no GitHub issue associated with this pull request

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h [hunk 1]
```diff
@@ -13,15 +13,15 @@ class WorkspaceProvider;
 
 class DLLExport EmptyWorkspaceNamePolicy {
 protected:
-  std::string getWorkspaceName(Mantid::API::IMDWorkspace_sptr) {
-    return "__EmptyWorkspaceNamePolicy";
-  }
+  const std::string &
+  getWorkspaceName(const Mantid::API::IMDWorkspace & /*workspace*/);
```
Comment (soininen) [inline Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h]: I thought we were trying to remove the comments from argument names in these cases, but I guess there is no project-wide policy on this.

File: Vates/VatesAPI/inc/MantidVatesAPI/vtkDataSetFactory.h [hunk 2]
```diff
@@ -148,13 +138,41 @@ class DLLExport vtkDataSetFactory
...
+   Try to cast it to the specified IMDType and then run checks based on the
+   non-integrated dimensionality.
+   The latter checks are only run if the factory is set to apply these checks.
+   @param  workspace : workspace to cast.
+   @param  bExactMatch : run an exact match on non-integarated dimensionality if
+   TRUE, otherwise is less than or equal to ExpectedDimensions.
+   @return  correctly cast shared pointer or an empty shared pointer if cast or
+   checks fail.
+   */
+
```
Comment (soininen) [inline Vates/VatesAPI/inc/MantidVatesAPI/vtkDataSetFactory.h]: Usually, there is no empty line here.

File: Vates/VatesAPI/src/PresenterFactories.cpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#include "MantidVatesAPI/PresenterFactories.h"
+
+namespace Mantid {
+namespace VATES {
+
+const std::string &EmptyWorkspaceNamePolicy::getWorkspaceName(
+    const Mantid::API::IMDWorkspace & /*workspace*/) {
+  static std::string name{"__EmptyWorkspaceNamePolicy"};
+  return name;
```
Comment (soininen) [inline Vates/VatesAPI/src/PresenterFactories.cpp]: Why not just `return "__EmptyWorkspaceNamePolicy"`?

**Review by soininen — CHANGES_REQUESTED** [review]

File: Vates/VatesAPI/src/PresenterFactories.cpp (hunk 3 above)
Comment (quantumsteve) [inline Vates/VatesAPI/src/PresenterFactories.cpp]: One can't return a non-static local variable by reference.

File: Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h (hunk 1 above)
Comment (quantumsteve) [inline Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h]: @mantidproject/tech-steering How should we mark unused arguments? Do we have or want to set a standard? 

FWIW I prefer commenting out the name vs. using the `UNUSED_ARG` macro or leaving it blank.

File: Vates/VatesAPI/src/PresenterFactories.cpp (hunk 3 above)
Comment (soininen) [inline Vates/VatesAPI/src/PresenterFactories.cpp]: Yes, indeed one can't.

File: Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h (hunk 1 above)
Comment (soininen) [inline Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h]: I have been using `UNUSED_ARG` but I have no strong feelings either way.

**Review by soininen — APPROVED** [review]

**Comment by soininen** [comment]
Nice tidying here. I don't think the unused function argument discussion above is a blocking issue, so by all means :shipit:

File: Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h (hunk 1 above)
Comment (martyngigg) [inline Vates/VatesAPI/inc/MantidVatesAPI/PresenterFactories.h]: @quantumsteve Does Doxygen cope with the commented out names?

My preference would be `UNUSED_ARG` as it's in more active use (I think) but I don't have strong feelings either way.

**Pull request merged**